### DISCUSSION TITLE
Intermittent CefSharp.core.Runime.dll could not be loaded

### DISCUSSION THREAD

**Discussion by cyberfizz** [cefsharp/CefSharp#3656] [General]
We have an intermittent issue with error: System.IO.FileLoadException: A procedure imported by 'CefSharp.Core.Runtime.dll'

It is a webform application (CefSharp v91.1.211) and we have added the bits for AnyCPU support, as well as adding in to the project file (VS2019):
```
<Target Name="CefSharpCopyFilesAfterNugetRestoreAnyCPU" AfterTargets="Build" Condition="'$(NuGetProjectStyle)' != 'PackageReference' AND !Exists('$(OutDir)$(CefSharpTargetDirAnyCpu32)libcef.dll') AND '$(CefSharpBuildAction)' != 'NoAction' AND '$(CefSharpPlatformTarget)' == 'AnyCPU'">
    <Message Importance="high" Text="CefSharp - Files were not copied by MSBuild after Nuget install/restore! Please close and re-open $(MSBuildProjectFile)." />
  </Target>
```

What's strange is if i do a build in VS it gives me the error. If i delete all the files, make a change in VS (I presume force it to compile again), then it works. As soon as i delete and rebuild though it goes back to same error. The only way to get it to work that i have found is by changing the code e.g. add a string etc.

Our code is using the bits from the Minimal Example and the resolver, with NoInlining on the LoadApp section:
```
  [STAThread]
        static void Main(string[] args)
        {
           
            try
            {
                AppDomain.CurrentDomain.AssemblyResolve += Resolver;
                LoadApp(args);
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.ToString());
                
            }

        }


private static Assembly Resolver(object sender, ResolveEventArgs args)
        {
            if (args.Name.StartsWith("CefSharp.Core.Runtime"))
            {
                string assemblyName = args.Name.Split(new[] { ',' }, 2)[0] + ".dll";
                string archSpecificPath = Path.Combine(AppDomain.CurrentDomain.SetupInformation.ApplicationBase,
                                                       Environment.Is64BitProcess ? "x64" : "x86",
                                                       assemblyName);

                if (File.Exists(archSpecificPath))
                {
                    return System.Reflection.Assembly.LoadFile(archSpecificPath);
                }
                else
                {
                    return null;
                }

            }
            return null;
        }     
```

Any ideas what is going on?

**Comment by amaitland** [comment]
> as well as adding in to the project file (VS2019)

Added what exactly? How does the CefSharpCopyFilesAfterNugetRestoreAnyCPU fit in exactly? 

> Whilst it is no longer necessary to close Visual Studio for the references to show up (dotnet/project-system#4158) the runtime dependencies, libcef.dll etc still aren't copied until the project is closed/reopened when using packages.config style projects. Using PackageReference works correctly and is the recommended option. Please note you can use PackageReference with older non SDK Style projects (ones that would typically have used packages.config) under VS2017/2019.

As per https://github.com/cefsharp/CefSharp/issues/3319 you'll need to close and reopen the project after a Nuget restore on clean checkout. Switching to PackageReference is recommend.

**Reply by amaitland** [reply]
https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference

**Reply by cyberfizz** [reply]
wow, thank you for the quick reply. I was reading something about the not copying of packages and thought the CefSharpCopyFilesAfterNugetRestoreAnyCPU  bit was relevant. 

I will try to change it over now to package reference. Thanks again. It was weird how it didnt always fully build, so as you say the packagereference should help.

**Reply by cyberfizz** [reply]
> https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference

This appears to have fixed it. Thanks again, i had never come across that before so every day is a school day.
